kyo-combinators/shared/src/main/scala/kyo/Combinators.scala

```diff
@@ -9,57 +9,130 @@ import scala.reflect.ClassTag
 
 extension [A, S](effect: A < S)
 
+    /** Performs this computation and then the next one, discarding the result of this computation.
+      *
+      * @param next
+      *   The computation to perform after this one
+      * @return
+      *   A computation that produces the result of `next`
+      */
     @targetName("zipRight")
     def *>[A1, S1](next: => A1 < S1)(using Frame): A1 < (S & S1) =
         effect.map(_ => next)
 
+    /** Performs this computation and then the next one, discarding the result of the next computation.
+      *
+      * @param next
+      *   The computation to perform after this one
+      * @return
+      *   A computation that produces the result of this computation
+      */
     @targetName("zipLeft")
     def <*[A1, S1](next: => A1 < S1)(using Frame): A < (S & S1) =
         effect.map(e => next.as(e))
 
+    /** Performs this computation and then the next one, returning both results as a tuple.
+      *
+      * @param next
+      *   The computation to perform after this one
+      * @return
+      *   A computation that produces a tuple of both results
+      */
     @targetName("zip")
     def <*>[A1, S1](next: => A1 < S1)(using Frame): (A, A1) < (S & S1) =
         effect.map(e => next.map(n => (e, n)))
 
+    /** Performs this computation, discards its result, and returns the given value.
+      *
+      * @param value
+      *   The value to return
+      * @return
+      *   A computation that produces the given value
+      */
     inline def as[A1, S1](value: => A1 < S1)(using inline frame: Frame): A1 < (S & S1) =
```

Comment:
I wonder if this should just accept pure values.

Reply:
Agreed. I didn't change the methods taking effectful inputs but that's the pattern in the rest of the codebase. Making the places where `map` happens more explicit I think makes the code easier to reason about, even if it introduces more syntax overhead. It'd be nice to hear @johnhungerford's thoughts on this.

Reply:
I do think it should be consistent with the rest of the codebase on this point. I'm kind of torn however: it's a really cool thing about Kyo that effects are perfectly composable and there's no cost to lifting a value to a "pure" kyo (A < Any). If there's a performance cost to the map, though, that would be a reason not to. But is that the case? Or are you concerned that the method signatures are more confusing?

Reply:
There is a performance cost but it's minimal, especially since it would not allocate if pure values are passed. I know I had highlighted "fluent APIs" as an advantage of Kyo's encoding but I've reconsidered this view. Making the monadic bindings more explicit helps to indicate where effects are coming from and simplifies method signatures. Although it introduces more syntax overhead, I believe it be beneficial for newcomers. It can be especially helpful when multiple effects are involved since it makes the flow of effects more transparent and clear to reason about.

Reply:
Seems reasonable. For contaminators, I think we will need per method decision making. `as` should probably accept non-kyo, but `tap` or other operators can accept effects.